ext/cl_khr_defined_builtin_kernels.asciidoc

```diff
@@ -99,222 +103,414 @@ accelerators while being able to synchronize and optimize data transfers at
+clCreateDefinedBuiltInKernelDescriptor.
+
+==== Sample Code ====
 
-4+| OpenCL C Semantics
-4+a|
 [source,c]
 ----
-__kernel void __khr_io_stream_in_blocking_uchar(uchar *output, size_t *num) {
-   while (*num) {
-       size_t num_read = *num;
-       __khr_io_stream_in_uchar(output, &num_read);
-       num -= num_read;
-       output += num_read;
-   }
+// TBD. Similarly in cl_qcom_ml_ops, tensors have type
+// (cl_channel_type) and a number of dimensions (rank) and dimension
+// sizes (shape). Difference over the cl_qcom_ml_ops is that the rank is
+// "unlimited".
+cl_tensor_desc_t lhs_tensor_desc = TBD;
+cl_tensor_desc_t rhs_tensor_desc = TBD;
+cl_tensor_desc_t res_tensor_desc = TBD;
+
+cl_dkb_attributes_matmul_v1 matmul_attrs = {
+  lhs_tensor_desc, rhs_tensor_desc, res_tensor_desc,
+  1, 0 // = Transpose lhs tensor
 }
-----
 
-|===
-
-==== Launching BiKs from the Device Side ====
-
-BiKs are primarily meant to be launched as kernel commands via host-side command queues.
-Optionally, they can be callable from device-side via
-`enqueue_kernel`: This capability can be queried on per BiK basis at compile-time in OpenCL C by checking for macro definitions which has the following naming convention: `cl_khr_bik_BUILTIN_KERNEL_NAME`. In case a BiK macro is defined, a kernel with a naming convention `__khr_BUILTIN_KERNEL_NAME()` can be enqueued by the program at device side as software-defined kernels.
+cl_dbk_mode_properties matmul_props = {
+  // Request a matmul implementation that meets this precision.
+  CL_DBK_PROPERTY_MAX_RELATIVE_ERROR, 100, // in ULPs.
+}
 
+cl_uint err;
+std::vector<cl_dbk_descriptor> kernel_descriptions;
+cl_dbk_descriptor matmul_desc =
+  clCreateDefinedBuiltInKernelDescriptor(
+  context, num_devices, device_list,
+  CL_DBK_MATMUL_V1, &matmul_attrs, &matmul_props, &err);
+
+} else if (err == CL_DBK_UNAVAILABLE) {
+  // Kernel attributes are valid but the kernel is not supported in at least
+  // one of the devices.
+} else if (err == CL_DBK_UNMET_MAX_RELATIVE_ERROR) {
+  // E.g. Kernel is supported but is not precise enough.
+} else if (err == CL_DBK_UNSUPPORTED_MODE_PROPERTY) {
+  // cl_dbk_mode_properties has a property not listed in the description of the
+  // defined built-in kernel.
+} else
+  kernel_descriptions.push_back(matmul_desc);
+
+...
+
+cl_program dbk_lib = clCreateProgramWithDefinedBuiltInKernels(
+  context, kernel_descriptions.size(), kernel_descriptors.data(), err);
+
+...
+
+cl_kernel matmul_kernel = clCreateDefinedBuiltinKernel(
+  dkb_lib, matmul_desc, err);
+
+// TBD: allocate space for tensors. Perhaps like cl_qcom_ml_ops: query
+// tensor sizes after the final program has been created or after
+// command buffer (with DBKs within) is finalized. Implementation
+// determines the optimal data layout (opaque to the application) for
+// the tensors based on their usage.  Application uses the tensor
+// sizes to create cl_mem buffers which are bound to the tensors.
+cl_tensor_t lhs_tensor = TBD;
```

Comment:
```suggestion
cl_tensor lhs_tensor = TBD;
```

Reply:
TBD after cl_khr_tensor draft.

Reply:
Actually the suggestion makes sense because it would be consistent with OpenCL's other API object types.